src/sentence.rs

```diff
@@ -0,0 +1,302 @@
+        ATerm,
+        UpperLower,
+        ClosePlus,
+        SpPlus,
+        STerm
+    }
+
+    #[derive(Clone, PartialEq, Eq)]
+    struct SentenceBreaksState(pub [StatePart; 4]);
+
+    const INITIAL_STATE: SentenceBreaksState = SentenceBreaksState([
+        StatePart::Sot,
+        StatePart::Sot,
+        StatePart::Sot,
+        StatePart::Sot
+    ]);
+
+    pub struct SentenceBreaks<'a> {
+        pub string: &'a str,
+        pos: usize,
+        state: SentenceBreaksState
+    }
+
+    impl SentenceBreaksState {
+        fn next(&self, cat: SentenceCat) -> SentenceBreaksState {
+            let &SentenceBreaksState(parts) = self;
+            let parts = match (parts[3], cat) {
+                (StatePart::ClosePlus, SentenceCat::SC_Close) => parts,
+                (StatePart::SpPlus, SentenceCat::SC_Sp) => parts,
+                _ => [
+                    parts[1],
+                    parts[2],
+                    parts[3],
+                    match cat {
+                        SentenceCat::SC_CR => StatePart::CR,
+                        SentenceCat::SC_LF => StatePart::LF,
+                        SentenceCat::SC_Sep => StatePart::Sep,
+                        SentenceCat::SC_ATerm => StatePart::ATerm,
+                        SentenceCat::SC_Upper |
+                        SentenceCat::SC_Lower => StatePart::UpperLower,
+                        SentenceCat::SC_Close => StatePart::ClosePlus,
+                        SentenceCat::SC_Sp => StatePart::SpPlus,
+                        SentenceCat::SC_STerm => StatePart::STerm,
+                        _ => StatePart::Other
+                    }
+                ]
+            };
+            SentenceBreaksState(parts)
+        }
+
+        fn end(&self) -> SentenceBreaksState {
+            let &SentenceBreaksState(parts) = self;
+            SentenceBreaksState([
+                parts[1],
+                parts[2],
+                parts[3],
+                StatePart::Eot
+            ])
+        }
+
+        fn match1(&self, part: StatePart) -> bool {
+            let &SentenceBreaksState(parts) = self;
+            part == parts[3]
+        }
+
+        fn match2(&self, part1: StatePart, part2: StatePart) -> bool {
+            let &SentenceBreaksState(parts) = self;
+            part1 == parts[2] && part2 == parts[3]
+        }
+    }
+
+    fn match_sb8(state: &SentenceBreaksState, ahead: &str) -> bool {
+        let aterm_part = {
+            // ATerm Close* Sp*
+            let &SentenceBreaksState(parts) = state;
+            let mut idx = if parts[3] == StatePart::SpPlus { 2 } else { 3 };
+            if parts[idx] == StatePart::ClosePlus { idx -= 1 }
+            parts[idx]
+        };
+
+        if aterm_part == StatePart::ATerm {
+            use tables::sentence as se;
+
+            for next_char in ahead.chars() {
+                //( ¬(OLetter | Upper | Lower | ParaSep | SATerm) )* Lower
+                match se::sentence_category(next_char) {
```

Comment:
this can be quadratic for cases where you're in the midst of the `Close*Sp*`. Worth caching the lookup.

Doesn't have to be done in this PR though (but please file an issue once this merges)